### PR TITLE
MultiDataSource feature merge

### PR DESCRIPTION
**Pull request by kristenTian**
### Description
Merge [multidata source feature branch]( https://github.com/opensearch-project/OpenSearch-Dashboards/tree/feature/multi-datasource) into main
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
    - [ ] `yarn test:jest`
    - [ ] `yarn test:jest_integration`
    - [ ] `yarn test:ftr`
- [ ] New functionality has been documented.
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: .github/workflows/build_and_test_workflow.yml [hunk 1]
```diff
@@ -6,9 +6,9 @@ name: Build and test
 # trigger on every commit push and PR for all branches except feature branches and pushes for backport branches
 on:
   push:
-    branches: [ '**', '!feature/**', '!backport/**' ]
+    branches: ['**', '!backport/**']
```
Comment (zhongnansu) [inline .github/workflows/build_and_test_workflow.yml:9]: we can revert this change that we made to the feature branch.

File: package.json [hunk 2]
```diff
@@ -110,6 +110,7 @@
     ]
   },
   "dependencies": {
+    "@aws-crypto/client-node": "^3.1.1",
```
Comment (zhongnansu) [inline package.json:113]: could you include the `yarn.lock` change? I assume since we added new dependency, yarn.lock is supposed to be updated

File: .github/workflows/build_and_test_workflow.yml (hunk 1 above)
Comment (kristenTian) [reply .github/workflows/build_and_test_workflow.yml:9]: will do

**Comment by kavilla** [comment]
DCO check seems to be failing and i think it needs an update snapshot

File: src/plugins/credential_management/public/components/edit_credential_wizard/edit_credential_wizard.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+          iconType,
+        },
+      ]);
+    }
+  };
+
+  if (credential) {
+    return <EditCredential credential={credential} />;
+  } else {
+    return <h1>Credential not found!</h1>;
```
Comment (AMoo-Miki) [inline src/plugins/credential_management/public/components/edit_credential_wizard/edit_credential_wizard.tsx:84]: "Credentials" ?

Comment (AMoo-Miki) [reply src/plugins/credential_management/public/components/edit_credential_wizard/edit_credential_wizard.tsx:84]: Or "No credentials stored" ?

File: src/plugins/credential_management/public/components/utils.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+          if (a.sort < b.sort) {
+            return -1;
+          }
+
+          if (a.sort > b.sort) {
+            return 1;
+          }
+
+          return 0;
+        })
```
Comment (AMoo-Miki) [inline src/plugins/credential_management/public/components/utils.ts:42]: We should use `localeCompare` here or even better use `Intl.Collator`.

File: src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts [hunk 5]
```diff
@@ -355,12 +357,14 @@ export class IndexPatternsService {
...
         type,
       },
+      references,
     } = savedObject;
 
     const parsedSourceFilters = sourceFilters ? JSON.parse(sourceFilters) : undefined;
     const parsedTypeMeta = typeMeta ? JSON.parse(typeMeta) : undefined;
     const parsedFieldFormatMap = fieldFormatMap ? JSON.parse(fieldFormatMap) : {};
     const parsedFields: FieldSpec[] = fields ? JSON.parse(fields) : [];
+    const dataSourceRef = !_.isEmpty(references) ? references[0] : undefined;
```
Comment (AMoo-Miki) [inline src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts:367]: `const dataSourceRef = Array.isArray(references) && references.length ? references[0] : undefined;`
or even
`const dataSourceRef = Array.isArray(references) && references[0] || undefined;`

File: src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts [hunk 6]
```diff
@@ -29,8 +29,8 @@
  */
 
 import { i18n } from '@osd/i18n';
+import _ from 'lodash';
```
Comment (AMoo-Miki) [inline src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts:32]: Please don't import "lodash".

File: src/plugins/data/server/index_patterns/fetcher/lib/opensearch_api.ts [hunk 7]
```diff
@@ -57,10 +57,20 @@ export interface IndexAliasResponse {
  *  @return {Promise<IndexAliasResponse>}
  */
 export async function callIndexAliasApi(
-  callCluster: LegacyAPICaller,
+  callCluster: LegacyAPICaller | OpenSearchClient,
   indices: string[] | string
 ): Promise<IndicesAliasResponse> {
   try {
+    if ('transport' in callCluster) {
```
Comment (AMoo-Miki) [inline src/plugins/data/server/index_patterns/fetcher/lib/opensearch_api.ts:64]: A comment here would help developers know the significance of having `callCluster.transport`

File: src/plugins/data_source/README.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# data_source
+
+A OpenSearch Dashboards plugin
+
+This plugin introduce OpenSearch data source into OpenSearch Dashboards, and provides related functions to connect to OpenSearch data sources.
```
Comment (AMoo-Miki) [inline src/plugins/data_source/README.md:5]: "introduces"

File: src/plugins/data_source/README.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# data_source
+
+A OpenSearch Dashboards plugin
```
Comment (AMoo-Miki) [inline src/plugins/data_source/README.md:3]: "An"

File: src/plugins/data_source/server/client/client_pool.ts [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+  start() {}
+
+  // close all clients in the pool
+  async stop() {
+    if (this.isClosed) {
+      return;
+    }
+    this.isClosed = true;
+    await Promise.all(this.cache!.values().map((client) => client.close()));
```
Comment (AMoo-Miki) [inline src/plugins/data_source/server/client/client_pool.ts:74]: What if any of those `client.close()` calls fail? Shouldn't we move the `his.isClosed = true;` after the successful closing of all?

**Review by AMoo-Miki — CHANGES_REQUESTED** [review]

File: src/plugins/data/server/index_patterns/fetcher/lib/opensearch_api.ts (hunk 7 above)
Comment (kristenTian) [reply src/plugins/data/server/index_patterns/fetcher/lib/opensearch_api.ts:64]: Will update, also adding the migration task for support data client with legacy client.

Comment (zhongnansu) [reply src/plugins/data/server/index_patterns/fetcher/lib/opensearch_api.ts:64]: we have to pending PR to add support for legacy client. After that PR is in, this will be removed
https://github.com/opensearch-project/OpenSearch-Dashboards/issues/2133

File: package.json (hunk 2 above)
Comment (zhongnansu) [reply package.json:113]: > Yarn resolves these issues around versioning and non-determinism by using lock files and an install algorithm that is deterministic and reliable. These lock files lock the installed dependencies to a specific version and ensure that every install results in the exact same file structure in node_modules across all machines.

https://stackoverflow.com/questions/39990017/should-i-commit-the-yarn-lock-file-and-what-is-it-for

Comment (kristenTian) [reply package.json:113]: sure, didn't know this is a convention here.

Comment (kristenTian) [reply package.json:113]: nvm, mani already added yarn lock in one PR

File: src/plugins/credential_management/README.md [hunk 11]
```diff
@@ -0,0 +1,11 @@
+# credentialManagement
+
+A OpenSearch Dashboards plugin
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/credential_management/README.md]: We should have deleted this plugin to save review effort, please ignore all file under credential_management path.

File: src/plugins/credential_management/public/components/common/text_content/text_content.ts [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    defaultMessage: 'Delete',
+  }
+);
+
+export const credentialEditPageAuthType = i18n.translate(
+  'credentialManagement.textContent.credentialEditPageAuthType',
+  {
+    defaultMessage: 'Username & password',
+  }
+);
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/credential_management/public/components/common/text_content/text_content.ts]: FYI @mpabba3003 in case in data source have the same issue

File: src/plugins/credential_management/public/components/common/components/create_credential_form/create_credential_form.tsx [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+  /* Validations */
+
+  isFormValid = () => {
+    const validationByField: CredentialFormValidation = {
+      title: [],
+      description: [],
+      username: [],
+      password: [],
+    };
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/credential_management/public/components/common/components/create_credential_form/create_credential_form.tsx]: FYI @mpabba3003 in case in data source have the same issue

File: src/plugins/data_source/server/client/client_pool.ts (hunk 10 above)
Comment (kristenTian) [reply src/plugins/data_source/server/client/client_pool.ts:74]: @zhongnansu please let me know if there is something else, otherwise I will go with the update

File: src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts (hunk 6 above)
Comment (kristenTian) [reply src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts:32]: updated to import method instead

**Comment by kristenTian** [comment]
> DCO check seems to be failing and i think it needs an update snapshot

updated

File: src/plugins/data_source/server/client/client_pool.ts (hunk 10 above)
Comment (zhongnansu) [reply src/plugins/data_source/server/client/client_pool.ts]: moving `this.isClosed=true` after Promise makes sense to me. Thanks for update

File: src/plugins/credential_management/public/components/common/components/create_credential_form/create_credential_form.tsx [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+  }
+
+  componentDidUpdate(
+    prevProps: Readonly<CredentialFormProps>,
+    prevState: Readonly<CredentialFormState>,
+    snapshot?: any
+  ) {
+    if (prevProps.callSubmit !== this.props.callSubmit) {
+      this.onClickSubmitForm();
+    }
```
[the remark this answers is not in the document]
Comment (mpabba3003) [reply src/plugins/credential_management/public/components/common/components/create_credential_form/create_credential_form.tsx]: This module has been removed

File: src/plugins/credential_management/public/components/create_button/create_button.tsx [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+interface Props {
+  history: History;
+}
+
+export const CreateButton = ({ history }: Props) => {
+  return (
+    <EuiButton
+      data-test-subj="createCredentialButton"
+      fill={true}
+      onClick={() => history.push('/create')}
```
[the remark this answers is not in the document]
Comment (mpabba3003) [reply src/plugins/credential_management/public/components/create_button/create_button.tsx]: I did not see any button debounce on click

File: src/plugins/credential_management/public/components/edit_credential_wizard/components/edit_credential.tsx [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+          <EuiModal onClose={closeModal}>
+            <EuiModalHeader>
+              <EuiModalHeaderTitle>
+                <h1>Update password</h1>
+              </EuiModalHeaderTitle>
+            </EuiModalHeader>
+
+            <EuiModalBody>
+              <EuiFieldPassword
+                placeholder="Passord field(focus)"
```
[the remark this answers is not in the document]
Comment (mpabba3003) [reply src/plugins/credential_management/public/components/edit_credential_wizard/components/edit_credential.tsx]: This module has been deleted. I will ensure i18n for other texts

File: src/plugins/credential_management/public/components/edit_credential_wizard/edit_credential_wizard.tsx (hunk 3 above)
Comment (mpabba3003) [reply src/plugins/credential_management/public/components/edit_credential_wizard/edit_credential_wizard.tsx]: This module has been removed

File: src/plugins/credential_management/public/components/common/text_content/text_content.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (mpabba3003) [reply src/plugins/credential_management/public/components/common/text_content/text_content.ts]: Thanks ... but nope

File: src/plugins/credential_management/public/components/utils.ts (hunk 4 above)
Comment (mpabba3003) [reply src/plugins/credential_management/public/components/utils.ts]: This module has been delete &  for data source management .. I have removed sort method as table is taking care of sort

File: src/plugins/data_source/server/cryptography/cryptography_client.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import {
+  buildClient,
+  CommitmentPolicy,
+  RawAesKeyringNode,
+  RawAesWrappingSuiteIdentifier,
+} from '@aws-crypto/client-node';
```
Comment (AMoo-Miki) [inline src/plugins/data_source/server/cryptography/cryptography_client.ts:11]: Why did we use this library over node's built-in [Crypto](https://nodejs.org/docs/latest-v14.x/api/crypto.html)?

**Review by AMoo-Miki — CHANGES_REQUESTED** [review]

File: src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+        color="danger"
+        iconType="trash"
+        onClick={() => {
+          setConfirmDeleteVisible(true);
+        }}
+        data-test-subj="deleteDataSourceConnections"
+        disabled={selectedDataSources.length === 0}
+      >
+        Delete {selectedDataSources.length || ''} connection
+        {selectedDataSources.length >= 2 ? 's' : ''}
```
[the remark this answers is not in the document]
Comment (mpabba3003) [reply src/plugins/data_source_management/public/components/data_source_table/data_source_table.tsx:120]: Nope this is how it looks
Screen Shot 2022-09-16 at 8 51 28 AM
Screen Shot 2022-09-16 at 8 51 36 AM
Screen Shot 2022-09-16 at 8 53 31 AM

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_data_source/components/header/header.scss [hunk 19]
```diff
@@ -0,0 +1,3 @@
+.dataSourceRadioHelperText {
+  text-indent: 24px;
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_data_source/components/header/header.scss:2]: synced offline, `px` is okay here due to elastic ui library is using the same unit all over the place

File: src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_data_source/components/header/header.tsx [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+            id="indexPatternManagement.createIndexPattern.stepDataSourceHeader"
+            defaultMessage="Step {currentStepNumber} of {totalStepNumber}: Configure data source"
+            values={{ currentStepNumber, totalStepNumber }}
+          />
+        </h2>
+      </EuiTitle>
+      <EuiText>
+        <FormattedMessage
+          id="indexPatternManagement.createIndexPattern.stepDataSourceLabel"
+          defaultMessage="Please pick the data source -- within which to configure index patterns."
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/index_pattern_management/public/components/create_index_pattern_wizard/components/step_data_source/components/header/header.tsx:74]: Agree, the text content on pages still pending copy from UX input, might be addressing all copy related in another PR depending when official copy will be provided.

File: src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts [hunk 21]
```diff
@@ -355,12 +357,14 @@ export class IndexPatternsService {
...
         type,
       },
+      references,
     } = savedObject;
 
     const parsedSourceFilters = sourceFilters ? JSON.parse(sourceFilters) : undefined;
     const parsedTypeMeta = typeMeta ? JSON.parse(typeMeta) : undefined;
     const parsedFieldFormatMap = fieldFormatMap ? JSON.parse(fieldFormatMap) : {};
     const parsedFields: FieldSpec[] = fields ? JSON.parse(fields) : [];
+    const dataSourceRef = !isEmpty(references) ? references[0] : undefined;
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data/common/index_patterns/index_patterns/index_patterns.ts:367]: Thanks for sharing this in very detailed explanation.  I think I heavily relied on SavedObject existing functionalities. --That a `references`  field will always present in the object, and in form or an empty/non-empty array.

File: src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts [hunk 22]
```diff
@@ -0,0 +1,273 @@
...
+        throw SavedObjectsErrorHelpers.createBadRequestError(`Invalid auth type: '${type}'`);
+    }
+  }
+
+  private async validateAndUpdatePartialAttributes<T = unknown>(attributes: T) {
+    const { auth, endpoint } = attributes;
+
+    if (endpoint) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        `Update data source endpoint is not supported`
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts:167]: updated

File: src/plugins/data_source/server/plugin.ts [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+  }
+
+  private getAuditMessage(request: OpenSearchDashboardsRequest, dataSourceId: string) {
+    const rawRequest = ensureRawRequest(request);
+    const remoteAddress = rawRequest?.info?.remoteAddress;
+    const xForwardFor = request.headers['x-forwarded-for'];
+
+    return xForwardFor
+      ? `${remoteAddress} attempted accessing through ${xForwardFor} on data source: ${dataSourceId}`
+      : `${remoteAddress} attempted accessing on data source: ${dataSourceId}`;
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data_source/server/plugin.ts:166]: Summarizing the idea - Forwarded is a more recommended source than x-forward-for due to it's better reliability. However not all proxy will include Forwarded in header. Therefore only pick it when exists.

File: src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts [hunk 24]
```diff
@@ -0,0 +1,273 @@
...
+          return attributes;
+        }
+      default:
+        throw SavedObjectsErrorHelpers.createBadRequestError(`Invalid credentials type: '${type}'`);
+    }
+  }
+
+  private validateAttributes<T = unknown>(attributes: T) {
+    const { title, endpoint, auth } = attributes;
+    if (!title) {
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts:200]: This is a good call. Just tested with creating visualizations though, the name could be just spaces. O_O.  But this will leave it in a broken state, so I say we shouldn't allow only spaces here.

File: src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts [hunk 25]
```diff
@@ -0,0 +1,273 @@
...
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'attribute "title" required for "data source" saved object'
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data_source/server/saved_objects/data_source_saved_objects_client_wrapper.ts:202]: Applied and updated error message

File: src/plugins/data_source_management/public/components/utils.ts [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+  await Promise.all(
+    selectedDataSources.map(async (selectedDataSource) => {
+      await deleteDataSourceById(selectedDataSource.id, savedObjectsClient);
+    })
+  );
+}
+
+export const isValidUrl = (endpoint: string) => {
+  try {
+    return Boolean(new URL(endpoint));
```
[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data_source_management/public/components/utils.ts:84]: added validation, we only support http & https right now

File: src/plugins/data_source/server/cryptography/cryptography_client.ts (hunk 17 above)
Comment (kristenTian) [reply src/plugins/data_source/server/cryptography/cryptography_client.ts:11]: This is related to the issue created by noCharger here:  [[MD] Research on encryption / decryption strategies · Issue #1756 · opensearch-project/OpenSearch-Dashboards](https://github.com/opensearch-project/OpenSearch-Dashboards/issues/1756) . in which will add more reasons about why pick AWS crypto solution.

To summary, The concern majorly to introduce aws sdk as a dependency while we could be using node native library. And the reason to still go with aws crypto is due to:
1. It provide an envelope encryption feature which is leveraged by current implementation: https://docs.aws.amazon.com/encryption-sdk/latest/developer-guide/concepts.html#envelope-encryption
2. For later plans integration with KMS could leverage the library 

Also one more question if we could add this dependency as a plugin dependency, but due to multidata source modifies base concepts including index pattern. The whole feature won't be easily implemented as a plugin.

File: src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts [hunk 27]
```diff
@@ -0,0 +1,74 @@
...
+      description: '',
+      endpoint: '',
+      auth: {
+        type: AuthType.UsernamePasswordType,
+        credentials: {
+          username: '',
+          password: '',
+        },
+      },
+    };
```
Comment (AMoo-Miki) [inline src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts:31]: We should add some tests for when this structure is not exactly maintained. For example, when `auth` or `credentials` prop is missing.

Comment (mpabba3003) [reply src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts:31]: Unit tests are included. Auth & Credentials are always expected to exist on types CreateDataSourceState & EditDataSourceState .. Without those the component will error out and would not even invoke validation methods.

File: src/plugins/data_source_management/public/components/validation/datasource_form_validation.ts [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+  /* Endpoint Validation */
+  if (!isValidUrl(formValues.endpoint)) {
+    validationByField.endpoint.push(dataSourceValidationEndpointNotValid);
+    formErrorMessages.push(dataSourceValidationEndpointNotValid);
+  }
+
+  /* Credential Validation */
+
+  /* Username & Password */
+  if (formValues.auth.type === AuthType.UsernamePasswordType) {
```
[the remark this answers is not in the document]
Comment (mpabba3003) [reply src/plugins/data_source_management/public/components/validation/datasource_form_validation.ts:77]: Auth & Credentials are always expected to exist on types CreateDataSourceState & EditDataSourceState .. Without those the component will error out and would not even invoke validation methods.

These methods are called only from Create/edit data source components ands these components should hold all the properties in its state.

[the remark this answers is not in the document]
Comment (kristenTian) [reply src/plugins/data_source_management/public/components/validation/datasource_form_validation.ts:77]: Given that this is a util function, and take input dealing with user input, I agree adding the optional chaining here, for the sake of safety for future evolution.

File: src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts (hunk 27 above)
Comment (kristenTian) [reply src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts:31]: We can add this in a separate PR.

Comment (AMoo-Miki) [reply src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts:31]: @mpabba3003 you mentioned "Without those the component will error out and would not even invoke validation methods"... a test should make sure that that happens, which is what I am pointing at.

Comment (mpabba3003) [reply src/plugins/data_source_management/public/components/validation/datasource_form_validation.test.ts:31]: oh ok.. I will write tests for that scenario

**Review by AMoo-Miki — APPROVED** [review]
Thank you!

File: src/plugins/data_source/server/cryptography/cryptography_client.ts (hunk 17 above)
Comment (AMoo-Miki) [reply src/plugins/data_source/server/cryptography/cryptography_client.ts:11]: @kristenTian @noCharger Let's capture this on #1756 as well.

Comment (noCharger) [reply src/plugins/data_source/server/cryptography/cryptography_client.ts:11]: @AMoo-Miki Thanks for approval, I am working on it.

@kristenTian  Thanks for summarize the offline discussion, I am gonna add more details in #1756
